### PR TITLE
removeAll and retainAll for GdxArray

### PR DESCRIPTION
**Pull request by cypherdare**
For #226.

### CODE REVIEW

File: collections/src/test/kotlin/ktx/collections/arraysTest.kt [hunk 1]
```diff
@@ -364,6 +364,25 @@ class ArraysTest {
...
+
+  @Test
+  fun `should retain elements from existing GdxArray`() {
+    val array = GdxArray.with(1, 2, 3, 4, 5)
+
+    array.retainAll { it % 2 == 1 }
+
+    assertEquals(GdxArray.with(1, 3, 5), array)
+  }
+
```
Comment (czyzby) [inline collections/src/test/kotlin/ktx/collections/arraysTest.kt]: Add tests for edge cases: empty arrays, no elements retained/removed.

File: collections/README.md [hunk 2]
```diff
@@ -32,7 +32,8 @@ chained.
 - Missing `addAll` and `removeAll` methods for arrays and iterables were added.
 - `iterate` method allows to iterate over collection's elements, while providing reference to `MutableInterator`. Can be
 used to easily remove collection elements during iteration.
-- `map`, `filter`, `flatten` and `flatMap` methods that work like methods in Kotlin stdlib but return `GdxArray`.
+- `removeAll` and `retainAll` higher order functions that work like functions in Kotlin stdlib
```
Comment (czyzby) [inline collections/README.md]: Could you also update the `CHANGELOG.md` file?

File: collections/src/main/kotlin/ktx/collections/arrays.kt [hunk 3]
```diff
@@ -225,6 +225,40 @@ inline fun <Type, R : Comparable<R>> GdxArray<out Type>.sortByDescending(crossin
   if (size > 1) this.sort(compareByDescending(selector))
 }
 
+/**
+ * Removes elements from the array that satisfy the predicate.
+ */
+inline fun <Type> GdxArray<Type>.removeAll(predicate: (Type) -> Boolean) {
+  var currentWriteIndex = 0
+  for (i in 0 until size){
```
Comment (czyzby) [inline collections/src/main/kotlin/ktx/collections/arrays.kt]: I don't think this file was formatted with KTX formatter.

**Comment by cypherdare** [comment]
I changes `this[]` to `items[]` to skip the unnecessary bounds check, since Array already provides direct access to the backing array for this sort of thing. Kind of trivial, but might as well.

What would you think about making overloads of these functions as well as `filter` that let you pass a pool to return removed items to? I could see myself using that a lot. Could be a separate PR.

File: collections/src/test/kotlin/ktx/collections/arraysTest.kt (hunk 1 above)
Comment (cypherdare) [reply collections/src/test/kotlin/ktx/collections/arraysTest.kt]: Also added cases for all elements retained/removed.

**Comment by czyzby** [comment]
> What would you think about making overloads of these functions as well as filter that let you pass a pool to return removed items to? I could see myself using that a lot. Could be a separate PR.

Seems like this could be simply returned by the functions, but I get how this is pretty situational and an extra parameter would be more efficient and customizable. I think you should submit a separate PR for that.

**Comment by cypherdare** [comment]
> > What would you think about making overloads of these functions as well as filter that let you pass a pool to return removed items to? I could see myself using that a lot. Could be a separate PR.
> 
> Seems like this could be simply returned by the functions, but I get how this is pretty situational and an extra parameter would be more efficient and customizable. I think you should submit a separate PR for that.

The idea arose as I realized I'd rarely be using these methods for what I'm usually doing with in-place removal, which is tracking many game objects that expire on a regular basis, usually particles. So the only thing I ever do with expired objects is either lose them to the GC or free them to a pool.

If we return all removed objects, we are allocating an array, which halfway defeats the purpose of removing in-place, and in many cases users would be ignoring the returned values anyway.

If we provide a parameter for a second lambda that is a handler for removed items, the syntax gets a little clunky for something that ~probably~ would only be used to return items to a pool (my opinion/guess).

So my vote is for just a pool. I'm questioning whether it should be a nullable optional parameter, or full overloads of the functions should be written, since they're inline and already kind of long.
